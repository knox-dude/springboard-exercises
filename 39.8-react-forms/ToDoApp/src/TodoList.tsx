import { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";

// Define interface for Todo
interface TodoProps {
  id: string;
  text: string;
}

function TodoList() {
  const [todos, setTodos] = useState<TodoProps[]>([]);

  const removeTodo = (id: string) => {
    setTodos(todos.filter((t) => t.id !== id));
  }

  const addTodo = (text: string, id: string) => {
    setTodos([...todos, {id, text}]);
  }

  const todoComponents = todos.map((todo: TodoProps) => (
    <Todo 
      key={todo.id} 
      id={todo.id} 
      text={todo.text} 
      remove={removeTodo}
    />
  ));

  return (
    <div>
      <NewTodoForm addTodo={addTodo} />
      <ul>
        {todoComponents}
      </ul>
    </div>
  )
}

export default TodoList;
