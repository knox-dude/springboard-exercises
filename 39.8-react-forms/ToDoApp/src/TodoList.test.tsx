import { render, screen, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

describe("TodoList", () => {

  //smoke test
  it("renders without crashing", () => {
    render(<TodoList />);
  });

  //snapshot test
  it("matches snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("can add a todo", () => {
    render(<TodoList />);
    const inputElement = screen.getByPlaceholderText('Enter todo...');
    const addButtonElement = screen.getByText('Add Todo');
    fireEvent.change(inputElement, { target: { value: 'Test Todo' } });
    fireEvent.click(addButtonElement);
  
    expect(screen.getByText('Test Todo')).toBeInTheDocument();
  });

  it("can remove a todo", () => {
    render(<TodoList />);
    // first add task so we can remove it
    const inputElement = screen.getByPlaceholderText('Enter todo...');
    const addButtonElement = screen.getByText('Add Todo');
    fireEvent.change(inputElement, { target: { value: 'Test Todo' } });
    fireEvent.click(addButtonElement);

    expect(screen.getByText('Test Todo')).toBeInTheDocument();
    expect(screen.getByText('Remove')).toBeInTheDocument();

    // remove the task
    const removeButtonElement = screen.getByText('Remove');
    fireEvent.click(removeButtonElement);

    expect(screen.queryByText('Test Todo')).toBeNull();
  });

});
