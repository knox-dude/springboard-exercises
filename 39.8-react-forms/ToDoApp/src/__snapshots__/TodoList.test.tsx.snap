// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        placeholder="Enter todo..."
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
